test/as-thrift.js

```diff
@@ -130,6 +130,93 @@ allocCluster.test('send and receive a notOk', {
     });
 });
 
+allocCluster.test('as=thrift send supports shouldApplicationRetry', {
+    numPeers: 3
+}, function t(cluster, assert) {
+    var tchannelAsThrift = makeTChannelThriftServer(cluster, {
+        notOkResponse: true,
+        succeedSecondTime: true,
+        servers: 2
+    });
+
+    var client = cluster.channels[2].subChannels.server;
+
+    tchannelAsThrift.send(client.request({
+        serviceName: 'server',
+        hasNoParent: true,
+        shouldApplicationRetry: shouldRetry
+    }), 'Chamber::echo', null, {
+        value: 10
+    }, function onResponse(err, res) {
+        assert.ifError(err);
+
+        assert.ok(res.ok);
+        assert.equal(res.body, 10);
+        assert.equal(res.headers.as, 'thrift');
+
+        assert.end();
+    });
+
+    function shouldRetry(req, res, retry, done) {
+        tchannelAsThrift.parseException(req, res, onException);
+
+        function onException(err, info) {
+            // Failed to parse thrift response
+            if (err) {
+                return done(err);
+            }
+
+            assert.equal(info.exception.value, 10);
+            assert.equal(info.exception.message, 'No echo');
+            assert.equal(info.typeName, 'noEcho');
+
+            if (info.typeName === 'noEcho') {
+                return retry();
+            }
+
+            done();
+        }
+    }
+});
+
+allocCluster.test('as=thrift request supports shouldApplicationRetry', {
+    numPeers: 3
+}, function t(cluster, assert) {
+    var tchannelAsThrift = makeTChannelThriftServer(cluster, {
+        notOkResponse: true,
+        succeedSecondTime: true,
+        servers: 2
+    });
+
+    tchannelAsThrift.request({
+        serviceName: 'server',
+        hasNoParent: true,
+        shouldThriftRetry: shouldRetry
+    }).send('Chamber::echo', null, {
+        value: 10
+    }, function onResponse(err, res) {
+        assert.ifError(err);
+
+        assert.ok(res.ok);
+        assert.equal(res.body, 10);
+        assert.equal(res.headers.as, 'thrift');
+
+        assert.end();
+    });
+
+    function shouldRetry(req, res, info, retry, done) {
+        assert.equal(info.exception.value, 10);
+        assert.equal(info.exception.message, 'No echo');
+        assert.equal(info.typeName, 'noEcho');
```

Comment:
We have an opportunity to make this not weird, I think. Maybe accept (res) return a bool? "res" being the result provided for onResponse at the Thrift layer?